### PR TITLE
cups vm tests: fix race condition, add more tests

### PR DESCRIPTION
**Pull request by Yarny0**
## Description of changes

Motivated by https://github.com/NixOS/nixpkgs/pull/337748#issuecomment-2313621353 , this adds

*   `cups-pdf` to cups' `passthru.tests`
*   two additional cups vm tests that ensure it works even with `listenAddresses = []`

While developing the new tests, I hit a rare race condition in the existing vm tests, which is also fixed as part of this pull request.  Details are explained in the respective commit message.

Notifying maintainers @domenkozar @matthewbauer and inspirator @risicle

Note: Although the [inspiring bug](https://github.com/NixOS/nixpkgs/pull/337748) occured on NixOS 24.05, I don't think this should be backported, as it's quite unlikely that such a bug happens again during the remaining lifetime of NixOS 24.05.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable ( **including added tests** ):
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

## Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

(nothing)

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/tests/printing.nix [hunk 1]
```diff
@@ -54,16 +56,16 @@ import ./make-test-python.nix (
...
+        client.wait_for_unit("ensure-printers.service")
 
     assert "scheduler is running" in client.succeed("lpstat -r")
 
     with subtest("UNIX socket is used for connections"):
         assert "/var/run/cups/cups.sock" in client.succeed("lpstat -H")
 
     with subtest("HTTP server is available too"):
-        client.succeed("curl --fail http://localhost:631/")
+        ${pkgs.lib.optionalString listenTcp ''client.succeed("curl --fail http://localhost:631/")''}
```
Comment (Aleksanaa) [inline nixos/tests/printing.nix:68]:
```suggestion
        ${lib.optionalString listenTcp ''client.succeed("curl --fail http://localhost:631/")''}
```

Comment (Yarny0) [reply nixos/tests/printing.nix]: Done.  However, that requires to `inherit (pkgs) lib;` before.

**Comment by Yarny0** [comment]
Thanks @Aleksanaa, I applied your suggestion.  I also rebased onto current `nixos-unstable`, and re-ran all affected tests (still build/pass).

**Comment by l0b0** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 338193`

**Review by l0b0 — APPROVED** [review]

**Comment by l0b0** [comment]
Tests pass on x64 Linux, thanks! Might need to exclude them from Darwin, though, based on the build failure.

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**